src/epcsaft/eos/permittivity.rs

```diff
@@ -0,0 +1,275 @@
+        state: &StateHD<D>,
+        dipole_scaling: &f64,
+        polarizability_scaling: &f64,
+        correlation_integral_parameter: &f64,
+    ) -> Self {
+        // reciprocal thermodynamic temperature
+        let boltzmann = 1.380649e-23;
+        let beta = (state.temperature * boltzmann).recip();
+
+        // Density
+        // let total_moles = state.moles.sum();
+        let density = state.moles.mapv(|n| n / state.volume).sum();
+
+        // dipole density y -> scaled dipole density y_star
+        let y_star = density * (beta * *dipole_scaling * 1e-19 + *polarizability_scaling * 3.) * 4.
+            / 9.
+            * PI;
+
+        // correlation integral
+        let correlation_integral = ((-y_star).exp() - 1.0) * *correlation_integral_parameter + 1.0;
+
+        // dielectric constan
+        let permittivity_pure = y_star
+            * 3.0
+            * (y_star.powi(2) * (correlation_integral * (17. / 16.) - 1.0) + y_star + 1.0)
+            + 1.0;
+
+        Self {
+            permittivity: permittivity_pure,
+        }
+    }
+
+    pub fn from_experimental_data(
+        data: &[Vec<(f64, f64)>],
+        temperature: D,
+        molefracs: &Array1<D>,
+    ) -> Self {
+        let permittivity = data
+            .iter()
+            .enumerate()
+            .map(|(i, d)| Self::interpolate(d.to_vec(), temperature).permittivity * molefracs[i])
+            .sum();
+        Self { permittivity }
+    }
+
+    pub fn from_perturbation_theory(
+        state: &StateHD<D>,
+        dipole_scaling: &[f64],
+        polarizability_scaling: &[f64],
+        correlation_integral_parameter: &[f64],
+        comp: &Array1<usize>,
+    ) -> Self {
+        //let nsolvent = comp.len();
+        // reciprocal thermodynamic temperature
+        let boltzmann = 1.380649e-23;
+        let beta = (state.temperature * boltzmann).recip();
+        // Determine scaled dipole density and correlation integral parameter of the mixture
+        let mut y_star = D::zero();
+        let mut correlation_integral_parameter_mixture = D::zero();
+        for i in comp.iter() {
+            let rho_i = state.partial_density[*i];
+            let x_i = state.molefracs[*i];
+
+            y_star +=
+                rho_i * (beta * dipole_scaling[*i] * 1e-19 + polarizability_scaling[*i] * 3.) * 4.
+                    / 9.
+                    * PI;
+            correlation_integral_parameter_mixture += x_i * correlation_integral_parameter[*i];
+        }
+
+        // correlation integral
+        let correlation_integral =
+            ((-y_star).exp() - 1.0) * correlation_integral_parameter_mixture + 1.0;
+
+        // permittivity
+        let permittivity = y_star
+            * 3.0
+            * (y_star.powi(2) * (correlation_integral * (17. / 16.) - 1.0) + y_star + 1.0)
+            + 1.0;
+
+        Self { permittivity }
+    }
+
+    pub fn interpolate(interpolation_points: Vec<(f64, f64)>, temperature: D) -> Self {
```

Comment:
Thanks a lot! I've changed this now to:
```/// Structure: &[(temperature, epsilon)]
    /// Assume ordered by temperature
    /// and temperatures are all finite.
    pub fn interpolate(interpolation_points: &[(f64, f64)], temperature: D) -> Self {
        // find index where temperature could be inserted
        let i = interpolation_points.binary_search_by(|&(ti, _)| {
            ti.partial_cmp(&temperature.re())
                .expect("Unexpected value for temperature in interpolation points.")
        });
        // if the result is OK we don't need to interpolate ...
        if let Ok(i) = i {
            return Self {
                permittivity: D::one() * interpolation_points[i].1,
            };
        }

        // if not, we can unwrap safely:
        let i = i.unwrap_err();
        let n = interpolation_points.len();

        // check cases:
        // 0.   : below lowest temperature
        // >= n : above highest temperature
        // else : regular interpolation

        let (l, u) = match i {
            0 => (interpolation_points[0], interpolation_points[1]),
            i if i >= n => (interpolation_points[n - 2], interpolation_points[n - 1]),
            _ => (interpolation_points[i - 1], interpolation_points[i]),
        };
        let permittivity_pure = (temperature - l.0) / (u.0 - l.0) * (u.1 - l.1) + l.1;

        Self {
            permittivity: permittivity_pure,
        }
    }

Reply:
Now my question: "if the result is OK we don't need to interpolate ...", what do we do with the derivative in that case. Of course, it's not well-defined at that point because it could be having the gradient "towards" the lower or the higher point. But if we do it like above, we would just get rid of it, which probably doesn't seem so right...

Reply:
`let i = i.unwrap_or_else(|i| i)` instead of `let i = i.unwrap_err();`